jdbc-handler/src/main/java/org/apache/hive/storage/jdbc/dao/HiveDatabaseAccessor.java

```diff
@@ -56,4 +57,19 @@ protected String addLimitToQuery(String sql, int limit) {
     }
     return sql + " LIMIT " + limit;
   }
+
+  @Override
+  protected List<String> getColNamesFromRS(ResultSet rs) throws Exception {
+    List<String> columnNames = super.getColNamesFromRS(rs);
+    return columnNames.stream()
+        .map(c -> {
+          int lastIndex = c.lastIndexOf(".");
+          return lastIndex == -1 ? c : c.substring(lastIndex + 1); })
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  protected String getMetaDataQuery(String sql) {
+    return addLimitToQuery(sql, 0);
```

Comment:
wont limit 0 return an empty set? why is this needed?

Reply:
The override method only cares about the query's result metadata, 
https://github.com/apache/hive/blob/a838ca2760799708c465051904b96cab4ae628d1/jdbc-handler/src/main/java/org/apache/hive/storage/jdbc/dao/GenericJdbcDatabaseAccessor.java#L128

An empty result set is enough for this purpose, no need to execute the whole query.

Reply:
Cant we just call addLimitToQuery(sql, 0) directly? Why add another method for it?

Reply:
The default implement of getMetaDataQuery is actually adding `limit 1` to the`sql`, 
https://github.com/apache/hive/blob/a838ca2760799708c465051904b96cab4ae628d1/jdbc-handler/src/main/java/org/apache/hive/storage/jdbc/dao/GenericJdbcDatabaseAccessor.java#L176-L178

I guess in some db vendors, `limit 0` will get an empty metadata of the result, that's why we cannot call `addLimitToQuery(sql, 0)` directly on all kinds of dbs.

Reply:
If my understanding is correct, this method only returns the query string, rather than actual result set. This method just relays the output string from another method, which seems like an extra method call to get the Query string.
calling `addLimitToQuery(sql, 0)` directly also should work fine in this case.
WDYT?

Reply:
This is where the method `getMetaDataQuery` takes place:
https://github.com/apache/hive/blob/master/jdbc-handler/src/main/java/org/apache/hive/storage/jdbc/dao/GenericJdbcDatabaseAccessor.java#L74-L109
`GenericJdbcDatabaseAccessor` is the parent class of all jdbc sources. In this example, `HiveDatabaseAccessor` is a child of `GenericJdbcDatabaseAccessor` and override some methods the `GenericJdbcDatabaseAccessor` provided, `getMetaDataQuery` is one of the such method.

I'm not sure how to add `addLimitToQuery(sql, 0)` directly unless we change this line:
https://github.com/apache/hive/blob/master/jdbc-handler/src/main/java/org/apache/hive/storage/jdbc/dao/GenericJdbcDatabaseAccessor.java#L86
but we couldn't assume the sql after `addLimitToQuery(sql, 0)` would return the correct column metadata on different kinds of sources.

Reply:
@akshat0395 Does this explanation answer your question? limit 1 returns a row but we are just looking for row metadata from the result set. 
This method is being overridden in this class because we want to use a limit 0 vs limit 1 that the parent class uses. Are you suggesting we change the GenericJdbcDatabaseAccessor.addLimitToQuery() ?

Reply:
@dengzhhu653 @nrg4878 Thanks for bringing the attention back on this, apologies for delay got caught up with other things.
Yes, the suggestion is to change GenericJdbcDatabaseAccessor.addLimitToQuery() and have a direct call for addLimitToQuery(sql, 0) in the base method itself. As you mentioned it only metadata and doesn't require limit 1.
I can understand why the current approach might have been take, in class `JethroDatabaseAccessor` we have done the same implementation of overriding the method and calling addLimitToQuery(sql, 0).
@dengzhhu653 are there any specific cases/sources where we would need limit 1 rather than limit 0 to get the metadata query?
If not, we should do this change at base itself.

Reply:
I tried on MySQL and Postgres, `limit 0` works for them. I wasn't applying the change in the base mainly because:
1. I couldn't figure out why it was on the original fix;
2. `limit 1` wouldn't introduce much overhead on these databases except Hive.

Reply:
For sake of simplicity for now, we can go ahead with the current changes, as you have already tested Mysql and postgres I don't think other datasources would have issues with limit 0 as well.
We can create a separate patch for refactor post some more testing around this and raise a patch then.
If that works for you please resolve the comment and move forward with the changes.
Thanks

Reply:
@akshat0395 Fair point about considering the change to limit 0 in the base code for all accessors. It think it would make sense if we are not doing anything with the first row of data. I also concur that it probably only adds significant overhead for hive.
I am good with the plan of forking it off into another jira. I would also like some additional work in confirming that we are pushing down the predicates to these DBs. 
@dengzhhu653  Can you create a follow up jira and merge this?

Reply:
Thanks @nrg4878 and @akshat0395, created HIVE-27583 for this discussion.